### PR TITLE
Add notebook to quickly generate and load embeddings in AlloyDB.

### PR DESCRIPTION
**Pull request by paulramsey**
## Add AlloyDB Embeddings Notebook

Note: Before submitting a pull request, please open an issue for discussion if you are not associated with Google.

## Checklist
- [X] I have followed [Sample Guidelines from AUTHORING_GUIDE.MD](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md)
- [X] README is updated to include [all relevant information](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#readme-file)
- [X] **Tests** pass:   `nox -s py-3.12` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#test-environment-setup))
- [X] **Lint** pass:   `nox -s lint` (see [Test Environment Setup](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/AUTHORING_GUIDE.md#test-environment-setup))
- [X] This sample adds a new sample directory, and I updated the [CODEOWNERS file](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/.github/CODEOWNERS) with the codeowners for this sample
- [X] Please **merge** this PR for me once it is approved

### CODE REVIEW

File: alloydb/notebooks/generate_batch_embeddings.ipynb [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+    {
+      "cell_type": "markdown",
+      "id": "WU0ruULySEbZ",
+      "metadata": {
+        "id": "WU0ruULySEbZ"
+      },
+      "source": [
+        "# Generate and store batch embeddings in AlloyDB\n",
+        "\n",
+        "This notebook walks you through generating Vertex AI embeddings for the AlloyDB database used by the [GenWealth Demo App](https://github.com/GoogleCloudPlatform/generative-ai/tree/main/gemini/sample-apps/genwealth). It dynamically builds a batch of text chunks to embed based on character length of the source data in order to get more results per inference, leading to much more efficient embeddings generation. It also uses psycopg to efficiently load the embeddings into AlloyDB after they are generated. These techniques can significantly speed up the process of generating large batches of embeddings and storing them in AlloyDB vs using the native embedding() function (about 6.5x faster based on limited testing)."
```
Comment (shweta345) [inline alloydb/notebooks/generate_batch_embeddings.ipynb:12]: This notebook shows you how to generate batch vector embeddings and store them in an AlloyDB database using the [GenWealth Demo App](https://github.com/GoogleCloudPlatform/generative-ai/tree/main/gemini/sample-apps/genwealth) as the sample dataset. 

With the steps listed here, you can dynamically build a batch of text chunks to embed based on character length of the source data in order to get more results per inference, leading to much more efficient embeddings generation. The process uses the `psycopg` library to efficiently load the embeddings into AlloyDB after they are generated. These techniques can significantly speed up the process of generating large batches of embeddings and storing them in AlloyDB vs using the native embedding() function (about 6.5x faster based on limited testing).

## Before you begin

* Download and set up the [GenWealth Demo App](https://github.com/GoogleCloudPlatform/generative-ai/tree/main/gemini/sample-apps/genwealth). 
* Running the steps in this notebook will incur Google Cloud charges. You may also be billed for Vertex AI API usages.

File: alloydb/notebooks/generate_batch_embeddings.ipynb [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+    {
+      "cell_type": "markdown",
+      "id": "jJGlsq4QSnvZ",
+      "metadata": {
+        "id": "jJGlsq4QSnvZ"
+      },
+      "source": [
+        "## Setup\n",
+        "\n",
+        "1. Install and import necessary packages."
```
Comment (shweta345) [inline alloydb/notebooks/generate_batch_embeddings.ipynb:24]: Can we remove the step numbering and make them H3 headings, such as ### Install and import required packages

File: alloydb/notebooks/generate_batch_embeddings.ipynb (hunk 1 above)
Comment (paulramsey) [reply alloydb/notebooks/generate_batch_embeddings.ipynb]: Thank you - Made these changes as requested.

File: alloydb/notebooks/generate_batch_embeddings.ipynb (hunk 2 above)
Comment (paulramsey) [reply alloydb/notebooks/generate_batch_embeddings.ipynb]: Made these changes as requested, and improved the documentation throughout.

**Review by kurtisvg — CHANGES_REQUESTED** [review]
Some quick issues I see while reviewing:

P0s:
- The code is not tested (which is [required to merge](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/main/CONTRIBUTING.md#contributing-a-patch))
- Versions installed w/ pip should be pinned to prevent issues with updates
- We should update to load data from a file -- either a file in this repo or a public GCS bucket to streamline implementation 

P1s: 
- User inputs should be streamlined to not rely on Secret Manager -- something like the "Set up alloyDB" section [here](https://github.com/googleapis/langchain-google-alloydb-pg-python/blob/main/samples/langchain_quick_start.ipynb) would be preferred 
- We should evaluate using asyncio -- much of this is blocked on I/O, may speed it up considerably to operate in parallel 
- Need to evaluate failure for vertex AI quota issues 


P2s:
- Should consider using a generator for a batching function (python best practice)
- We may also want to consider using the AlloyDB Python Connector here for mTLS + make execution in the codelab better

**Comment by glasnt** [comment]
Closed in favour of https://github.com/GoogleCloudPlatform/python-docs-samples/pull/12721

**Pull request closed without merging**